### DISCUSSION TITLE
.Net LaTeX projects

### DISCUSSION THREAD

**Discussion by charlesroddie** [verybadcat/CSharpMath#270] [General]
Over the last year we seem to have a resurrection of https://github.com/ForNeVeR/wpf-math , which is based on JMathTex , and now CSharpMath , which is based on https://github.com/kostub/iosMath .

This is very promising. I am very interested in what the project maintainers think about this, how different these two projects are, and how they might develop.

I started looking at wpf-math and have just begun a port to F#, the main interest being to simplify the code structure to the point where I can add features: LaTeX editing, SkiaSharp, and rendering text + formulas. But this is a very big task, and would end up creating yet another .Net LaTeX project!

I would be keen to talk with anyone interested in .Net LaTeX here or on Skype (my user id is my public email here).

@ForNeVeR @Happypig375

**Comment by Happypig375** [comment]
One good thing, one bad thing.

Bad thing first:

One concern that I have about using F# is that it is not supported under UWP (.NET Native).

I have tried using [this F# library](https://github.com/mathnet/mathnet-symbolics) on UWP half a year ago with no success. I tried everything from using the FSharp.Core Nuget package directly to decompiling the binary to C# and referencing it from the project. Either it fails to compile or throws System.TypeInitializationException at runtime.

I ended up calling [a JavaScript library](https://github.com/jiggzson/nerdamer) in a WebView.
So, sorry to say this, but I cannot use your new library if it is written in F#.

Now on to the good thing.

These two libraries can potentially be merged together.

Both internally use a DrawingContext.
The one in CSharpMath is not cross-platformy enough, with CoreText-like APIs which I could not map to Android/UWP. I don't know about WpfMath though, although it looks promising, and those XAML could be modified a bit to be [XAML Standard](https://github.com/Microsoft/xaml-standard) compliant.

We could have WPF and Xamarin.iOS supported together! Yay! :tada:

By the way, I am also tagging @verybadcat who you missed.

**Comment by ForNeVeR** [comment]
@Happypig375 re XAML: it isn't actually a big deal for WPF-Math. Our users may create their own XAML controls and use the rendering logic from the library. It's actually the renderer that matters, not XAML.

We're currently developing abstractions to decouple our library from WPF, but there's still more work to do for that.

From my brief look at CSharpMath, it seems that it is highly-configurable, and if you provide all the interface implementations for [`TypesettingContext`](https://github.com/verybadcat/CSharpMath/blob/b3f1a21b6579ff595b255c527e36dbb8f0c3197f/CSharpMath/FrontEnd/TypesettingContext.cs), it should work on any .NET Standard-compatible platform; that's brilliant.

Regarding the ways to combine WPF-Math and CSharpMath: at the moment I'm not sure. I can definitely see the CSharpMath potential (e.g. it seem to have implemented the features WPF-Math currently lacks, e.g. matrices) and I'll take a closer look at the implementation when I'll start working on similar features of WPF-Math. Although we'll need to think more about how to integrate our libraries into something bigger and better (if we eventually want to do it).

I'm open for discussion.

**Comment by charlesroddie** [comment]
> @Happypig375 One concern that I have about using F# is that it is not supported under UWP (.NET Native).

Yes unfortunately UWP support for F# is a problem and while it is hopefully temporary, Microsoft does seem to have [brushed it under the carpet](https://github.com/Microsoft/visualfsharp/issues/1096) for now @cartermp. Now that @dsyme is going to work with Xamarin things could change as the problem makes F# somewhat toxic for writing libraries intended for use with Xamarin.Forms. The remaining work on support seems much smaller than it was a couple of years ago, only that there is no-one doing this work.

The main advantage for [using F#](https://github.com/ForNeVeR/wpf-math/issues/111) would be expressing recursive tree structures and parsing LaTeX into these structures can make things _much_ easier than in non-functional languages. However if there is going to be a single C# project that is very well structured I will be happy to avoid multiplication of libraries and will try to contribute to that, albeit more slowly in C# than F#.

The iosMath project that CSharpMath is based on has support for an editor https://github.com/kostub/MathEditor and a port of that could be feasible.

**Comment by verybadcat** [comment]
I am glad to see this issue.

I had never noticed WPFMath until this issue came along. About different front ends, yes, my intention is that it should be possible to plug them in without altering the base project. The big things you need are a font and a MathTable. The former is what you are going to draw; the latter tells you how much space you need to allow around your glyphs. Presumably WPFMath contains the same information in some form or another.

**Comment by verybadcat** [comment]
Also, I don't think different languages in the core and front end projects should be a problem.

**Comment by Happypig375** [comment]
@charlesroddie
>only that there is no-one doing this work.

And there is a decent chance that it will never be done.
>The main advantage for using F# would be expressing recursive tree structures and parsing LaTeX into these structures can make things much easier than in non-functional languages

I have heard that record types are to be added in C# 8. This should hopefully solve this problem.
>However if there is going to be a single C# project that is very well structured I will be happy to avoid multiplication of libraries and will try to contribute to that, albeit more slowly in C# than F#.

Yup, record types should make it easier to structure a library well.

@verybadcat 

It does sound like that CSharpMath and WPFMath can be merged together as the information needed 
 and the methods to draw are very similar.

>Also, I don't think different languages in the core and front end projects should be a problem.

Yup. C# projects, F# projects, VB.NET projects etc. can all reference each other. But to distribute an entire library FSharp.Core is still needed to be referenced, which does not work on UWP, due to tail-calling MSIL etc.

**Comment by FoggyFinder** [comment]
So, as a conclusion after a week of silence. Do we going to see new F# library?

**Comment by Happypig375** [comment]
No definitive answer as of yet; @charlesroddie supports this idea and I am against it.

Btw, thank you for re-igniting this discussion.

**Comment by charlesroddie** [comment]
@FoggyFinder There are two C# projects which derive from two OO projects in other languages. So there is a lot of existing OO work. In this context it doesn't make sense to port to F# now, even if it is more inherently suitable. I did start to port wpf-math because I thought I would have to work on the internals in order to eventually implement a LaTeX editor. But CSharpMath has the possibility to port over MathEditor. So my currently strategy is to wait for these things to happen and help where I can. If over the next few months there are roadblocks or I if I have to understand the workings of the libraries to implement a feature, then an F# port will emerge.

**Comment by Happypig375** [comment]
I will try not to make that port happen ;)

**Comment by Happypig375** [comment]
Just found this today: https://assetstore.unity.com/packages/tools/gui/texdraw-51426

Someone is charging money for that.

**Comment by Happypig375** [comment]
One immediate downside of it is the sheer number of non-standard commands it uses.
@ForNeVeR Kinda reminds me of that person who still commit to the master branch of his fork...

**Comment by ForNeVeR** [comment]
I'm still not sure about what to consider a "LaTeX standard". It's hard to distinct between standard and non-standard stuff in today's TeX installations.

Regarding "that person": I believe that's a good man and we'll improve our project (and, well, the whole ecosystem of LaTeX-related projects) with help of people like that. Even if I have some difficulties communicating my requirements to people not familiar with git, that's my problem, and I will work harder to solve that.

**Comment by Happypig375** [comment]
By "LaTeX standard", I mean the documentation of it has to be at least searchable on the web.
For instance,

```
\frac{fraction}{with line}
\nfrac{fraction}{without line}
\lfrac{left}{fraction}
\rfrac{right}{fraction}
```
![screenshot_2018-09-01-19-03-43-601_com android chrome](https://user-images.githubusercontent.com/19922066/44945365-7cbddb00-ae1a-11e8-8eed-3b00f3400ad2.png)
No related information can be found about the last three aside from the reference for the asset.

**Comment by charlesroddie** [comment]
I think it should be a long-term goal to combine wpf-math and CSharpMath.

The place to start is probably the Atom type (called Atom in wpf-math and MathAtom in CSharpMath I believe). I think we should be able to use the same code for this. (Allowing for Atoms that cannot be rendered if one project doesn't support them yet.) Then we can share the LaTeX parser and lexer.

**Comment by Happypig375** [comment]
What name should the combined project be?

**Comment by charlesroddie** [comment]
I don't think that needs a decision now. A shared Atom project would have independent value even if the projects don't fully merge. Combining effort and ideas, simplifying the structure of the projects and reducing maintenance. @ForNeVeR @verybadcat @gsomix @alexreg does this make sense to you?

**Comment by alexreg** [comment]
@charlesroddie Agreed on the shared Atom structure, yes.

**Comment by ForNeVeR** [comment]
In principle, I approve the idea. But to give a meaningful answer, I'll need to invest some time into reading CSharpMath sources and researching of its' own Atom (or Atom-like) structures/types.

**Comment by Happypig375** [comment]
@ForNeVeR I will do the work. After all, CSharpMath's project structure is much bigger than wpf-math's. Please wait for my initial plan. (Will post here.)

**Comment by Happypig375** [comment]
## CSharpMath: road towards 0.1.0
- Subpixel rendering will be re-targeted to after the merger of CSharpMath and wpf-math. (Details below.)
- After fixing the bug that @FoggyFinder raised, 0.1.0 final will be released. (I can't do it right now as I am on a school trip to Taiwan.)

## The name of the new project
### Why?
> I don't think that needs a decision now.

I think it needs a decision these few days if not now.
1. The name is going to be immutable as it directly correlates to the project structure, and changing both will be a breaking change. An initial bad name will stick to the project. An example of a project not following this is [Typography.GlyphLayout](https://github.com/LayoutFarm/Typography/blob/c882af81ef893a2c431702c409e3594a4ab4fce6/Typography.GlyphLayout/GlyphLayout.cs). The project name is Typography.GlyphLayout but its namespace is Typography.TextLayout. It is hard to figure out where the classes are located without looking at the source.
2. A good name makes a good first impression and is easier to memorize. By enabling developers to memorize the name more easily, they will more likely to come back to the project after first sight. The more recurring developers, the more potential contributors. In the end, the project will be more active.
3. A good name will prevent static mindsets. Current names limit the scope of the projects. "CSharpMath" implies the project is only written in C# (discourages e.g. F# unit tests), and "wpf-math" implies the project is only in WPF (discourages porting to e.g. Avalonia). In the same way, WinForms, WPF and UWP are confined to Windows as implied in their names. By not including specific technologies in the name, the project will not be discouraged to include more technologies.

### What
I have brainstormed a few names. Here are my top picks, starting from my favorite:
#### MixiMath
1. "Mix" of CSharp"Math" and wpf-"math" (With correct ordering - "math" is at the end)
2. Name is short and concise
3. Searchable - Google only shows 8 results - all are irrelevant
4. Memorizable - It contains an alliteration!
5. Pronunciation unlikely to be incorrect as it is easy to pronounce.
#### FuseMath
1,2,5 still applicable
3,4 is not as great as MixiMath
Con: Brief stop in pronunciation between "Fuse" and "Math"
#### AtoMath
1: The "atoms" of CSharp"Math" and wpf-"math"
2,5 still applicable
3 is not as great as MixiMath
4: Memorizable - "Atom" is coined with "math"!
Con: "AtoMath" is already taken in GitHub. That means we won't be able to create an organization with such a name. We will have to use the not-so-fair collaborator system or use e.g. GitLab.
#### MathCore
1: Just like .NET "Core" is the new .NET platform, it is the new CSharp"Math" and wpf-"math".
2,5 still applicable
3 is just not applicable. "MathCore" is already used by other repositories.
4 is not present as well. "MathCore" seems like a conventional and uninnovative name.

### How
I am going to create an organization with the final name (MixiMath for example). There will be three repositories initially.
The main repository will be named MixiMath/MixiMath. It will contain the main library.
The second repository will be named MixiMath/Meta. It will contain design notes and moderation information (e.g. rules, contributing guide, warning/ban logs). It will be used for meta-discussion and discussion that do not correspond to other repositories.
Last but not least, the third repository will be named MixiMath/MixiMath.github.io. It will contain documentation and a github website through GitHub Pages.

The people invited will include (rough plan):

Team | People | Subteams
-|-|-
Core | @ForNeVeR, @Happypig375 | _(none)_
Advisors | @charlesroddie, @FoggyFinder, @alexreg, @gsomix, @verybadcat | Core
Collaborators | @B3zaleel | Advisors

Please do comment for improvements.

**Comment by charlesroddie** [comment]
There is so far no agreement on a merger of wpf-math and csharpmath but there is some agreement in principle about sharing code, specifically Atom/MathAtom, subject to investigating its feasibility.

We can start a new repo to put relevant code/issues/documentation/metaissues in but I would not start three. If we do this now I would suggest for a name AtomicMath if you want something that sounds good, but the namespace should be MathAtom or Atom. 

This repo will be for creating a shared MathAtom type and possibly a parser that converts LaTeX math into a MathAtom. This project will be successful if it produces a good MathAtom type that the maintainers of wpf-math and csharpmath want to use. After that, who knows.

The thing I am most interested is a github issue that we can start discussing the Atom type in. Given @Happypig375 's post I'll wait for a new repo to do that in.

**Comment by B3zaleel** [comment]
What about "*__Mathigrafia__*"? (from Greek Mathimatikí typografía aka mathematical typography).

**Comment by Happypig375** [comment]
@charlesroddie I'll wait for @ForNeVeR's response to feed my thought.
@B3zaleel Isn't that name a bit hard to remember?

**Comment by B3zaleel** [comment]
Not really

**Comment by alexreg** [comment]
It's a bit of a mouthful though, I reckon. I would suggest something like *MathType* or *MathScript*, but I think they're already taken by reasonably well-known software.

**Comment by ForNeVeR** [comment]
@Happypig375 alright, your plan makes sense for me.

I'd vote for FuseMath (it sounds cool IMHO), but other ideas such as MixiMath, AtoMath and Mathigrafia are good as well. As a main project author (at least for now), you decide.

About the organizational structure: I agree with @charlesroddie. For now, let's just start a one new repository and start experiments there. I tend to agree in a way with your ideas about future organizational structure, but now let's just start the programming already :)

(And please send us a repository link. I'll try to keep a close eye on your initial development, so I could explore the new Atom structure while you're still working on it.)

Later, when we'll be ready, I'm in principle okay to transfer WPF-Math repo to the new organization (as long as I keep full administrative control* over the said repo).

_\* Full administrative control is important to me not because I'm plotting something, but because I use it to set up CI and various automation stuff in my repositories; it's just required for my work as a maintainer. You know me, I'm a rational man and it shouldn't be a problem._

**Comment by alexreg** [comment]
Not a fan of FuseMath, to be honest.

**Comment by Happypig375** [comment]
@alexreg
Is it too edgy?

**Comment by alexreg** [comment]
@Happypig375 Perhaps yes. I'm just not sure I like the phonetic quality or that it gives it a good indication of what it's about (the "fuse"). I'm being picky though...

**Comment by Happypig375** [comment]
@alexreg The phonetic quality is what was noted above. "FuseMath" is not about its use, but its history: It started as an atomic "fusion" between the CSharpMath atoms and wpf-math atoms.

**Comment by alexreg** [comment]
Yeah, but that is kind of obscure. Most users will scratch their heads at that.

**Comment by Happypig375** [comment]
What about "MixiMath"?

**Comment by alexreg** [comment]
Better... still not 100% sure though.

**Comment by Happypig375** [comment]
What could be improved?

**Comment by verybadcat** [comment]
About merging the projects -- what would CSharpMath gain from a merger?

**Comment by alexreg** [comment]
The initial idea is just shared object hierarchy for atoms; nothing more, I think.

**Comment by Happypig375** [comment]
@verybadcat
With a limited number of contributors,
we can either:
1. Work on two separate projects and duplicate the effort of maintaining the code and developing new features; or
2. Work on one combined project, maintain half of the code, and develop new features once only.

I'd pick 2.

**Comment by Happypig375** [comment]
Let's conclude...

Name|Voters|Total Votes
------|-------|------------
MixiMath|@ForNeVeR (0.5), @alexreg, @Happypig375|2.5
FuseMath|@ForNeVeR, @Happypig375 (0.5), @FoggyFinder|2.5
AtomicMath|@charlesroddie, @FoggyFinder|2
Mathigrafia|@B3zaleel|1

A tie between MixiMath and FuseMath.

**Comment by FoggyFinder** [comment]
I like FuseMath it sounds really cool, AtomicMath also good name

**Comment by Happypig375** [comment]
Erm... Late vote, but I'll also count.

**Comment by charlesroddie** [comment]
We can go with FuseMath as the name for the current cooperation project. If the project ends up creating a complete solution usable by end-users and not only by wpf-math and csharpmath, it will need a new name as the fuse term is only meaningful to insiders (same goes for mix). I just created a MathAtom repo but we can use FuseMath instead. Do you want to create the repo @Happypig375 or shall I rename MathAtom?

**Comment by Happypig375** [comment]
Note: The name itself does not really need to be very meaningful aside from representing the company/project. It just has to sound good, be one word, so it can be memorable.
See: Google, Xamarin, Tencent...

**Comment by charlesroddie** [comment]
I created this yesterday: https://github.com/charlesroddie/MathAtom

Feel free to change the name.

Started to document the Atoms: https://github.com/charlesroddie/MathAtom/issues/2 . Need more info on the CSharpMath MathAtom as there are features that CSharpMath has that don't seem to be represented in https://github.com/verybadcat/CSharpMath/tree/master/CSharpMath/Atoms .

**Comment by verybadcat** [comment]
I agree that picking up contributors is a wonderful goal, and if someone wants to join the project, they should be encouraged and helped. But my feeling is, don't share code just to pick up contributors. That's just looking for trouble down the line, when a change to support one project screws up the other one. Good reasons to share code would be to include more platforms or to typeset things we otherwise couldn't typeset.

That said, the fact that I started this project should not make my opinion in any way dispositive.

**Comment by Happypig375** [comment]
That's why I would want to completely deprecate the old projects eventually. It's all or nothing.

By the way, please vote on a project name.

**Comment by ForNeVeR** [comment]
@verybadcat let's consider that a research project: it will help us to conclude if code sharing between .NET LaTeX projects is possible, and if it is, then at what degree. You aren't obliged to accept the "merge" or any "shared" code, but I think that's still an important research, even if it will eventually fail or be fruitless.

I think the strategic goals were already discussed in this thread: eventually we'd like to merge the best parts of CSharpMath and WPF-Math together, and that (theoretically, at least) should benefit all our users.

**Comment by verybadcat** [comment]
I'm all for research. And you don't have to persuade me. If you think I'm wrong, just do it. About the name, I don't think the fact of a merger is significant to the name. What about TexNet or MathDraw?

**Comment by Happypig375** [comment]
@verybadcat TexNet: See [3rd item of Why in this comment](https://github.com/verybadcat/CSharpMath/issues/4#issuecomment-425196488). A name like that would discourage supporting e.g. MathML, which I am planning for the future.

MathDraw: Google it and see the amount of name collisions.

**Comment by Happypig375** [comment]
The fact that the proposed names are inspired by the merge does not mean they are not good names. [See the What section of my comment](https://github.com/verybadcat/CSharpMath/issues/4#issuecomment-425196488). They are short, concise, memorable, special enough, and does not seem to be reused much elsewhere. The fact that they are inspired by the merge is just icing on the cake.

**Comment by charlesroddie** [comment]
MathPrinter or MathRender ?

**Comment by Happypig375** [comment]
Note: Searching "MathPainter" yields the CSharpMath.Rendering.MathPainter class on GitHub. LOL!

**Comment by alexreg** [comment]
MathScript? MathScript-net/MathScript.NET? So many good names are already taken, I'm often in favour of appending "-net" or ".NET".

**Comment by charlesroddie** [comment]
I like MathScript.NET.

**Comment by Happypig375** [comment]
The website "mathscript.net" is already taken, but I think the name itself is good enough to be behind MixiMath and FuseMath in my favorites list.

"MathDraw.NET", "MathPrint.NET" and "MathRender.NET" are available though.

New nomination: "MathDisplay" (does not collide with other projects even without ".NET"!)

MixiMath and FuseMath are still on my favorites because they look specially enough to be memorable. [See Point 2 of the Why section of my comment.](https://github.com/verybadcat/CSharpMath/issues/4#issuecomment-425196488)

**Comment by charlesroddie** [comment]
I like MathDisplay.

**Comment by alexreg** [comment]
I prefer MathScript.NET (we don't necessarily need the domain name). MathDraw and MathDisplay are also decent though. :-)

**Comment by Happypig375** [comment]
### Name comparison table

Name|No. of characters|No. of syllables|No. of Google results|No. of Google results with quotes|No GitHub name collision?|No GMail name collision?
-|-|-|-|-|-|-
MixiMath|8|3|8|8|✔️|✔️
FuseMath|8|2|33|37|✔️|✔️
AtoMath|7|3|2 630|2 940|❌|❌
MathCore|8|2|592 000|530 000|❌|❌
AtomicMath|10|4|1 760|1 640|✔️|❌
Mathigrafia|11|5|0|0|✔️|✔️
MathType|8|2|1 290 000|1 130 000|❌|❌
MathScript|10|2|59 500|52 500|✔️|❌
TexNet|6|2|369 000|311 000|✔️|❌
MathDraw|8|2|13 700|12 800|✔️|❌
MathPrinter|11|3|114|112|✔️|❌
MathRender|10|3|379|415|✔️|✔️
MathPainter|11|3|500|507|✔️|❌
MathDraw.NET|12|4|14 500|1|⍻|✔️
MathPrint.NET|13|4|37 200|3|⍻|✔️
MathRender.NET|14|5|417|0|⍻|✔️
MathDisplay|11|3|48 900|8 370|✔️|✔️

⍻ in "No GitHub collision?" means that hyphens will have to be used. Not beautiful.
Memorizability is not included as it is subjective. 

Welp, the rightmost column really takes a lot of names out of consideration.

### My current top picks (considering the above table)
1. MixiMath
2. FuseMath
3. MathRender
4. MathDisplay
5. Mathigrafia
6. MathDraw.NET, MathPrint.NET, MathRender.NET

**Comment by Happypig375** [comment]
By the way,
MathRender/MathDisplay -> "It renders/displays math!"
MixiMath/FuseMath -> "I don't know what it does. I'll check it out!"

If we constrain the name to "It renders/displays math!" only, users might exclude the possibility of the library being capable of other math-related features, e.g. math evaluation. (Yes, this is planned)

**Comment by B3zaleel** [comment]
Rendering and evaluation should be kept separate else the repo can become harder to maintain.

**Comment by charlesroddie** [comment]
> "I don't know what it does. I'll check it out!"

That's not how people behave. We should have a descriptive name.

> the library being capable of other math-related features, e.g. math evaluation

The current library is to do with layout and the atom is based on the laid out structure. Conversion to e.g. a Math.Net Symbolics expression to evaluate would still be to do with layout, since it is understanding a laid out expression. So most of the descriptive names would still work.

We could think about joining Math.Net. Math.Net Display/Type/... .

**Comment by alexreg** [comment]
> That's not how people behave. We should have a descriptive name.

Agreed.

I'm with Charles on the second point too.

**Comment by Happypig375** [comment]
@B3zaleel
Yea, keeping them separate is easier to maintain, to the point where the popular one gets all of the development attention, and the others bite the dust. See: Math.NET Numerics vs other Math.NET projects.

@charlesroddie, @alexreg 
https://novanym.com/blogs/blog/16463485-business-naming-six-reasons-why-you-shouldn-t-be-descriptive

TL;DR Stand out from the crowd.

About merging with Math.NET, I'm not sure about that. As mentioned above, Math.NET is kinda numerics-first, symbolics-second. I want to let the new project to be symbolics-first.

**Comment by alexreg** [comment]
@Happypig375 Fair point about Math.NET. About "standing out from the crowd"... not sure this is standing out for a good reason!

**Comment by Happypig375** [comment]
Why not?

**Comment by alexreg** [comment]
Well, obscurity...

**Comment by Happypig375** [comment]
As mentioned by the blog I linked, people aren't going to be able to search the new project by typing MathDraw/MathPrint/etc., since there would be so much results. Referencing the Name Comparison Table I posted, there are really lots of search results for the descriptive names. In contrast, MixiMath only has 8, FuseMath only has 33.

**Comment by ForNeVeR** [comment]
@Happypig375 I'm all for merging symbolic evaluation contributions, if any, with Math.NET's symbolics project. It's already somewhat widespread and well-known, and we don't have too much .NET math-based projects around to split our efforts, so I think we should better interoperate with other people.

As a side point, Math.NET maintainers don't mind F# (I've discussed it with them in scope of another project). They are very nice people, so probably you could discuss your plans with them (visit their Gitter chat if you like) and you'll come out with benefits for both your project and for Math.NET.

**Comment by Happypig375** [comment]
There's also [Symbolism](https://github.com/dharmatech/Symbolism) but it seems to be dead at the moment.

**Comment by alexreg** [comment]
Good to hear, @ForNeVeR! I am now leaning towards that. `Math.NET Display/Graphics/Draw` seems like a good name.

**Comment by ForNeVeR** [comment]
@alexreg well, I didn't mean that our hypothetical LaTeX-focused project should be a part of Math.NET, but...

Now, when you say that, that actually seems like a decent idea. Although I think that we'll need more certainty on our side before offering to join Math.NET. Let's keep this in mind but don't do any additional steps before we ship something usable.

**Comment by Happypig375** [comment]
Well, I'll tag @cdrnet for potential thoguhts on this.

**Comment by alexreg** [comment]
@ForNeVeR Fair enough.

**Comment by Happypig375** [comment]
@mathnet Would you consider us merging with you?

**Comment by cdrnet** [comment]
This projects sounds very interesting! I'd be happy to work together with you on this, but we'd need to make sure that I cannot block you, i.e. no task like creating a new release should depend on me (as there may be phases of time where I cannot respond quickly).

(btw: Math.NET Symbolics can also format its expressions in LaTeX)

**Comment by Happypig375** [comment]
So, let's vote.
A. Creative naming (MixiMath, FuseMath, etc.)
B. Descriptive naming (MathRender, MathDisplay, etc.)

I vote A.

**Comment by alexreg** [comment]
B for me.

**Comment by FoggyFinder** [comment]
Hard choice, but my vote for `A` too.

**Comment by charlesroddie** [comment]
B with a preference for MathDisplay.Net which could become Math.Net Display in future.

**Comment by ForNeVeR** [comment]
These days, I tend to use "creative" names for end-user projects (i.e. executables), but "technical" names for library projects, so they would be easier to distinguish for the programmers involved in end-user projects.

Consider yourself changing your work and entering a development of a math-based end-user solution. So, your first day at work, and you skim over the package lists to get a first look at what each of projects in your solution do. Probably it would be better for you to see

```
- MathDisplay.Net
- Json.Net
- Amazon.WpfIntegration
```

than

```
- GloryOfSauron
- Mighty.Rhinoceros
- =++Octarine++=
```

So, for a library project my vote goes for B.

**Comment by verybadcat** [comment]
Just thinking about libraries I use in my own work, they tend to be descriptively named. So I vote B.

**Comment by Happypig375** [comment]
A: 2
B: 4

I guess descriptive naming wins.

"MathDisplay.Net" seems to be the name that is the current consensus. Should we format the name as "MathDisplay.Net" or "MathDisplay.NET"? I prefer the latter as it aligns with "Math.NET" and ".NET" in general.

**Comment by alexreg** [comment]
> These days, I tend to use "creative" names for end-user projects (i.e. executables), but "technical" names for library projects, so they would be easier to distinguish for the programmers involved in end-user projects.

I follow this pattern too. It seems most appropriate given whom libraries and programs are targeted at respectively.

**Comment by Happypig375** [comment]
"MathDisplay.NET" it is then.

**Comment by mauler2025** [comment]
> Just found this today: https://assetstore.unity.com/packages/tools/gui/texdraw-51426
> 
> Someone is charging money for that.

@Happypig375 : Did the guy behind TeXDraw use some of CSharpMath libraries (parser, etc)?

**Comment by charlesroddie** [comment]
I don't believe @willnode is using CSharpMath or WpfMath in any way in his Unity package but he is welcome do to so and keep charging money as the licenses for both projects are MIT.

**Comment by mauler2025** [comment]
The source code of CSharpMath is not well documented compared to Wpf-Math. It is hard to digest what is going on behind the scene. No vertical spacing also makes it unnecessarily harder. :-)

**Comment by Happypig375** [comment]
@pstricks-fans What do you suggest to improve documentation? What do mean by vertical spacing?

**Comment by mauler2025** [comment]
> @pstricks-fans What do you suggest to improve documentation? What do mean by vertical spacing?

No empty lines between methods, no enough comments describing the used algorithm, etc.

Suggestion: Provide at least a diagram describing how the execution flows from TeX input to rendered output.

**Comment by Happypig375** [comment]
>No empty lines between methods

With CodeLens turned on, there are effectively empty lines between methods.
![image](https://user-images.githubusercontent.com/19922066/82775689-fd3c2f00-9e7a-11ea-8c66-324d2b3c25d1.png)

When empty lines are further added, there are effectively 2 lines between methods.
![image](https://user-images.githubusercontent.com/19922066/82775729-1e048480-9e7b-11ea-89ac-57027357f98a.png)

That's why empty lines tend not to be added between methods.

>no enough comments describing the used algorithm, etc.

>Suggestion: Provide at least a diagram describing how the execution flows from TeX input to rendered output.

Will add, thanks

**Comment by mauler2025** [comment]
It will be nice if you provide us with a description on the main page about how the underlying mechanism used to process the TeX input to render the corresponding output. Which classes can be categorized into the common code base  so they can be reused for various targets (Xamarin.Forms, Unity, Blazor web-assembly, and much more).  Thank you!

**Comment by Happypig375** [comment]
@pstricks-fans Added https://github.com/verybadcat/CSharpMath/blob/master/ReadMe.md#major-processes-of-drawing-latex

**Comment by mauler2025** [comment]
Thank you very much @Happypig375.

**Comment by yoshiask** [comment]
Is there any update on getting this project ready for Math.NET?

**Comment by Happypig375** [comment]
@yoshiask At this moment it is just an idea, no concrete plans yet. You can suggest a concrete plan though!
